### PR TITLE
new mesh capabilities in component caps

### PR DESCRIPTION
**Pull request by mvertens**
new mesh capabilities in component caps
This PR puts new mesh capability into each component nuopc cap and into the data model caps

- new mesh files are now being generated in the directory:
  $DIN_LOC_ROOT/share/meshes
  The process for generating these files is to determine th SCRIP grid description file that
  was used to generate the corresponding domain file and use the command
     mpirun -np 2 ESMF_Scrip2Unstruct <scrip_file> <mesh file> 0 ESMF

- each prognostic component now:
  - after its own initialization, the component reads in an ESMF mesh file
  - a new mesh is created using the read in mesh and a distgrid that
    is created from the global index space produced by the component
    initializaiton
      - for clm, the distgrid that is used to create the mesh has
        ocean points added to the end of each land gindex so that a
        global distgrid without holes can be sent back to the mediator
      - for cice, the distgrid that is used to create the mesh has any
        land block eliminated points added back into the the pe0 so that
        a global distgrid without holes can be sent back to the mediator
  - a check is done to determine if the mesh lats and lons are the same as
    the lats and lons generated internally by the component to some set tolerance
    level

- each data component now:
  - reads in the ESMF mesh and obtains its distgrid from the read in mesh
  - uses that distgrid to generate a gsmap for the component
  - passes that gsmap to the shr_strdata_init call - thereby eliminating
    the need for an mct rearranger in the data model nuopc cap.
  - a check is done to determine if the mesh lats and lons are the same as
    the lats and lons generated internally by the component to some set tolerance
    level
  - currently the component ggrid uses the domain lats and lons for the mapping
    of the streams to the component - but the mesh lats and lons to pass back to the
    mediator - THIS NEEDS TO BE ELIMINATED in the next update since it will change
    answers

- the data model shr_strdata_mod and shr_stream_mod have been modified as follows:
  - shr_strdata_init has been broken up into 3 subroutines:
      - shr_strdata_init_model_domain
      - shr_strdata_init_streams
      - shr_strdata_init_mapping
     this has enabled the capability to call these routines independently and
     not require the ordering imposed by the initial implementation
   - shr_dmodel_mod now can have a gsmap passed in directly in order to
     determine the component ggrid

- added new tests for DWAV and DLND

- notes on comparison to old baseline (/glade/scratch/mvertens/BASELINES/nov08)

  - The following tests showed a failure - but the only difference was fill pattern diffs
     ERS_Vnuopc_Ld5.T62_g16.GMOM.cheyenne_intel.mom-nuopc_cap
     ERS_Vnuopc_Ln5.f45_f45_mg37.I2000Clm50SpNuopc.cheyenne_intel.clm-nuopc_cap
     SMS_Vnuopc_Ld5.T62_g16.GMOM.cheyenne_intel.mom-nuopc_cap
     SMS_Vnuopc_Ld5.f19_g16.BMOM.cheyenne_intel.allactive-nuopc_cap_io
     SMS_Vnuopc_Ln5.f45_f45_mg37.I2000Clm50SpNuopc.cheyenne_intel.clm-nuopc_cap

  - The following tests differed in roundoff levels at the first timesteps
     ERS_Vnuopc_Ln5.f19_g17.F2000Nuopc.cheyenne_intel.cam-nuopc_cap
     SMS_Vnuopc_Ln5.f19_g17.F2000Nuopc.cheyenne_intel.cam-nuopc_cap
     ERS_Vnuopc_Ln9_N3.T31_g37_rx1.A.cheyenne_intel
     SMS_Vnuopc_Ld1_N3.T31_g37_rx1.A.cheyenne_intel

- new baselines generated in /glade/scratch/mvertens/BASELINES/nov12

Test suite: testlist_cmeps.xml
Test baseline: nov08 in /glade/scratch/mvertens/BASELINES
Test namelist changes: no
Test status: bit for bit and roundoff (see summary)
Fixes: None
User interface changes?: No
Update gh-pages html ?:N
Code review: jedwards

### CODE REVIEW

File: config/cesm/config_grids.xml [hunk 1]
```diff
@@ -1053,6 +1058,7 @@
       <file grid="ocnice"  mask="gx1v6">domain.ocn.0.47x0.63_gx1v6_090408.nc</file>
       <file grid="atm|lnd" mask="gx1v7">domain.lnd.fv0.47x0.63_gx1v7.180521.nc</file>
       <file grid="ocnice"  mask="gx1v7">domain.ocn.fv0.47x0.63_gx1v7.180521.nc</file>
+      <mesh driver="nuopc">$DIN_LOC_ROOT/share/meshes/fv0.47x0.63_141008_ESMFmesh.nc</mesh>
```
Comment (jedwards4b) [inline config/cesm/config_grids.xml:1061]: What is the significance of 141008?  It's not a date.   Do we need a new naming convention for these files or can we follow an established one?

File: config/cesm/config_grids.xml [hunk 2]
```diff
@@ -1121,30 +1129,32 @@
 
     <domain name="T62">
       <nx>192</nx>  <ny>96</ny>
-      <file grid="atm|lnd" mask="gx1v7">$DIN_LOC_ROOT/share/domains/domain.lnd.T62_gx1v7.151008.nc</file>
-      <file grid="atm|lnd" mask="gx1v6">$DIN_LOC_ROOT/share/domains/domain.lnd.T62_gx1v6.090320.nc</file>
-      <file grid="atm|lnd" mask="gx3v7">$DIN_LOC_ROOT/share/domains/domain.lnd.T62_gx3v7.090911.nc</file>
-      <file grid="atm|lnd" mask="tx1v1">$DIN_LOC_ROOT/share/domains/domain.lnd.T62_tx1v1.090122.nc</file>
+      <file grid="atm|lnd" mask="gx1v7"  >$DIN_LOC_ROOT/share/domains/domain.lnd.T62_gx1v7.151008.nc</file>
```
Comment (jedwards4b) [inline config/cesm/config_grids.xml:1132]: remove trailing whitespace.

File: config/cesm/machines/config_machines.xml [hunk 3]
```diff
@@ -332,7 +332,8 @@ This allows using a different mpirun command to launch unit tests
       <env name="MPI_USE_ARRAY"/>
     </environment_variables>
     <environment_variables comp_interface="nuopc">
-      <env name="ESMFMKFILE">/glade/u/home/dunlap/ESMF-INSTALL/8.0.0bs16/lib/libg/Linux.intel.64.mpich2.default/esmf.mk</env>
+      <!-- <env name="ESMFMKFILE">/glade/u/home/theurich/ESMF-INSTALLS/ESMF_fixedSMM/lib/esmf.mk</env> -->
```
Comment (jedwards4b) [inline config/cesm/machines/config_machines.xml]: remove commented code

File: config/xml_schemas/config_grids_v2.xsd [hunk 4]
```diff
@@ -86,13 +87,21 @@
         <xs:element ref="nx"/>
         <xs:element ref="ny"/>
         <xs:element ref="file" minOccurs="0" maxOccurs="unbounded"/>
+        <xs:element ref="mesh" minOccurs="0" maxOccurs="unbounded"/>
```
Comment (jedwards4b) [inline config/xml_schemas/config_grids_v2.xsd]: Shouldn't this be maxOccurs=1?   Can you have more than one mesh per domain?

File: src/components/data_comps/datm/nuopc/atm_comp_nuopc.F90 [hunk 5]
```diff
@@ -96,11 +89,12 @@ subroutine SetServices(gcomp, rc)
     integer, intent(out) :: rc
 
     ! local variables
+    integer :: dbrc
     character(len=*),parameter  :: subname=trim(modName)//':(SetServices) '
     !-------------------------------------------------------------------------------
 
     rc = ESMF_SUCCESS
-    if (dbug > 5) call ESMF_LogWrite(subname//' called', ESMF_LOGMSG_INFO, rc=dbrc)
+    call ESMF_LogWrite(subname//' called', ESMF_LOGMSG_INFO, rc=dbrc)
```
Comment (jedwards4b) [inline src/components/data_comps/datm/nuopc/atm_comp_nuopc.F90:97]: why did you remove dbug level?

**Comment by rsdunlapiv** [comment]
Submitted tests:
```
dunlap@cheyenne4:~/UFSCOMP.nov13/cime/scripts> qcmd -l walltime=04:00:00 -- ./create_test --xml-machine cheyenne --xml-testlist testlist_cmeps.xml --baseline-root /glade/scratch/mvertens/BASELINES --compare nov12
Submitting command to PBS using account P93300606:
    ./create_test --xml-machine cheyenne --xml-testlist testlist_cmeps.xml --baseline-root /glade/scratch/mvertens/BASELINES --compare nov12
```

File: src/drivers/nuopc/mediator/med.F90 [hunk 6]
```diff
@@ -475,18 +475,13 @@ subroutine InitializeP0(gcomp, importState, exportState, clock, rc)
 
     write(msgString,'(A,i6)') trim(subname)//' dbug_flag = ',dbug_flag
     call ESMF_LogWrite(trim(msgString), ESMF_LOGMSG_INFO, rc=dbrc)
-
-    if (dbug_flag > 5) then
-      call ESMF_LogWrite(trim(subname)//": called", ESMF_LOGMSG_INFO, rc=dbrc)
-    endif
+    call ESMF_LogWrite(trim(subname)//": called", ESMF_LOGMSG_INFO, rc=dbrc)
```
Comment (jedwards4b) [inline src/drivers/nuopc/mediator/med.F90:478]: why remove all the dbug_flags?   We need to agree on a way to do this that allows runtime control of these messages and implement it consistently across cime.

File: src/drivers/nuopc/shr/med_constants_mod.F90 [hunk 7]
```diff
@@ -37,6 +37,6 @@ module med_constants_mod
 
   !-----------------------------------------------------------------------------
   ! TODO: This is not a constant and should be moved elsewhere
-  integer :: med_constants_dbug_flag = 0
+  integer :: med_constants_dbug_flag = 2
```
Comment (jedwards4b) [inline src/drivers/nuopc/shr/med_constants_mod.F90]: set back to 0

**Review by jedwards4b — CHANGES_REQUESTED** [review]
Again - way too large for one PR.  It's very difficult to review.  Requested changes are minor issues.

File: config/cesm/config_grids.xml (hunk 1 above)
Comment (mvertens) [reply config/cesm/config_grids.xml:1061]: I will make the date change in a separate PR. The date in the filename should be the creation date.

File: config/xml_schemas/config_grids_v2.xsd (hunk 4 above)
Comment (mvertens) [reply config/xml_schemas/config_grids_v2.xsd]: Done.

File: config/cesm/machines/config_machines.xml (hunk 3 above)
Comment (mvertens) [reply config/cesm/machines/config_machines.xml]: Done.

File: src/drivers/nuopc/mediator/med.F90 (hunk 6 above)
Comment (mvertens) [reply src/drivers/nuopc/mediator/med.F90:478]: We talked about this a while ago - and we agreed that the levels should go away and we should just get debug output in the PET files in debug mode.

File: src/components/data_comps/datm/nuopc/atm_comp_nuopc.F90 (hunk 5 above)
Comment (mvertens) [reply src/components/data_comps/datm/nuopc/atm_comp_nuopc.F90:97]: See below.

**Review by jedwards4b — APPROVED** [review]

**Review by rsdunlapiv — APPROVED** [review]

**Comment by rsdunlapiv** [comment]
Generated new baselines: `/glade/scratch/dunlap/BASELINES/nov13`

Externals used:
```
s   ./cime
        clean sandbox, nuopc-cmeps --> origin/nuopc-cmeps
    ./components/cam
        clean sandbox, on cam1/branches/nuopc_cap/components/cam
    ./components/cice
        clean sandbox, on 86dcb4449d017097d455dcb74e6d2f71f469c5bd
    ./components/clm
        clean sandbox, on 16d8d769ed9f103b950c2fabe4781e5e4d28a178
    ./components/clm/src/fates
        clean sandbox, on fates_s1.8.1_a3.0.0
    ./components/clm/tools/PTCLM
        clean sandbox, on PTCLM2_180214
    ./components/mom
        clean sandbox, on 7ff64af1e4adf47c1d1c39c4323c1a3c0b5923d6
    ./components/mom/MOM6
        clean sandbox, on 4a5a5d87deb89ac876dca005fdb1b0046b779541
    ./components/mom/MOM6/pkg/MOM6_DA_hooks
        clean sandbox, on 6d8834ca8cf399f1a0d202239d72919907f6cd74
    ./components/mom/MOM6/pkg/geoKdTree
        clean sandbox, on a4670b9743c883d310d821eeac5b1f77f587b9d5
    ./components/mosart
        clean sandbox, on 77bc2aaf2b32e48601bcbf8e3b70527fad3db6e2
    ./libraries/CVMix
        clean sandbox, on v0.92-beta
    ./libraries/FMS
        clean sandbox, on 4a198e65401ba605e0fc44d1a2c0f4c9166dd6b4
    ./libraries/FMS/src
        clean sandbox, on warsaw_201803
```
